File: magda/agents/clip_agent.py
from typing import Dict, Any, List
import uuid
from .base import BaseAgent
import openai
import os
from dotenv import load_dotenv
from ..models import ClipResult, DAWCommand, AgentResponse

load_dotenv()

class ClipAgent(BaseAgent):
    """Agent responsible for handling clip creation operations using LLM."""
    
    def __init__(self):
        super().__init__()
        self.created_clips = {}
        self.client = openai.OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
    
    def can_handle(self, operation: str) -> bool:
        """Check if this agent can handle clip operations."""
        return operation.lower() in ['clip', 'create clip', 'add clip']
    
    def execute(self, operation: str, context: Dict[str, Any] | None = None) -> Dict[str, Any]:
        """Execute clip creation operation using LLM."""
        context = context or {}
        
        # Use LLM to parse clip operation and extract parameters
        clip_info = self._parse_clip_operation_with_llm(operation)
        
        # Get track information from context
        track_id = context.get("track_id")
        if not track_id:
            # Try to get track from context
            track_context = context.get("track")
            if track_context and "id" in track_context:
                track_id = track_context["id"]
        
        # Generate unique clip ID
        clip_id = str(uuid.uuid4())
        
        # Create clip result
        clip_result = ClipResult(
            id=clip_id,
            track_id=track_id or "unknown",
            start_bar=clip_info.get("start_bar", 1),
            end_bar=clip_info.get("end_bar", clip_info.get("start_bar", 1) + 4)
        )
        
        # Store clip for future reference
        self.created_clips[clip_id] = clip_result.dict()
        
        # Generate DAW command
        daw_command = self._generate_daw_command(clip_result)
        
        response = AgentResponse(
            result=clip_result.dict(),
            daw_command=daw_command,
            context=context
        )
        
        return response.dict()
    
    def _parse_clip_operation_with_llm(self, operation: str) -> Dict[str, Any]:
        """Use LLM to parse clip operation and extract parameters using Responses API."""
        
        instructions = """You are a clip creation specialist for a DAW system.
        Your job is to parse clip creation requests and extract the necessary parameters.
        
        Extract the following information:
        - start_bar: The starting bar number (default: 1)
        - end_bar: The ending bar number (default: start_bar + 4)
        - track_reference: Any reference to a specific track
        
        Return a JSON object with the extracted parameters following the provided schema."""
        
        try:
            response = self.client.responses.parse(
                model="gpt-4.1",
                instructions=instructions,
                input=operation,
                text_format=ClipResult,
                temperature=0.1
            )
            
            # The parse method returns the parsed object directly
            return response.output_parsed.dict()
            
        except Exception as e:
            print(f"Error parsing clip operation: {e}")
            return {}
    
    def _generate_daw_command(self, clip: ClipResult) -> str:
        """Generate DAW command string from clip result."""
        return f"clip(track:{clip.track_id}, start:{clip.start_bar}, end:{clip.end_bar})"
    
    def get_capabilities(self) -> List[str]:
        """Return list of operations this agent can handle."""
        return ["clip", "create clip", "add clip"]
    
    def get_clip_by_id(self, clip_id: str) -> Dict[str, Any]:
        """Get clip information by ID."""
        return self.created_clips.get(clip_id)
    
    def list_clips(self) -> List[Dict[str, Any]]:
        """List all created clips."""
        return list(self.created_clips.values())
    
    def get_clips_for_track(self, track_id: str) -> List[Dict[str, Any]]:
        """Get all clips for a specific track."""
        return [clip for clip in self.created_clips.values() if clip["track_id"] == track_id] 

File: cpp/src/agents/clip_agent.cpp
#include "magda_cpp/agents/clip_agent.h"
#include "prompt_loader.hpp"
#include <algorithm>
#include <iostream>
#include <sstream>

namespace magda {

ClipAgent::ClipAgent(const std::string& api_key)
    : BaseAgent("clip", api_key) {}

bool ClipAgent::canHandle(const std::string& operation) const {
    std::string op_lower = operation;
    std::transform(op_lower.begin(), op_lower.end(), op_lower.begin(), ::tolower);

    return op_lower.find("clip") != std::string::npos ||
           op_lower.find("region") != std::string::npos ||
           op_lower.find("recording") != std::string::npos ||
           op_lower.find("audio clip") != std::string::npos ||
           op_lower.find("midi clip") != std::string::npos;
}

AgentResponse ClipAgent::execute(const std::string& operation, const nlohmann::json& context) {
    try {
        // Load shared prompt
        SharedResources resources;
        std::string prompt = resources.getClipAgentPrompt();

        // Create schema for clip parameters
        auto schema = JsonSchemaBuilder()
            .type("object")
            .title("Clip Parameters")
            .description("Parameters for creating clips in a DAW")
            .property("start_bar", JsonSchemaBuilder()
                .type("integer")
                .description("Starting bar number"))
            .property("end_bar", JsonSchemaBuilder()
                .type("integer")
                .description("Ending bar number"))
            .property("start_time", JsonSchemaBuilder()
                .type("number")
                .description("Start time in seconds"))
            .property("duration", JsonSchemaBuilder()
                .type("number")
                .description("Clip duration in seconds"))
            .property("track_name", JsonSchemaBuilder()
                .type("string")
                .description("Target track name"))
            .required({"start_bar", "end_bar", "start_time", "duration", "track_name"})
            .additionalProperties(false);

        // Parse clip operation using base class method
        auto clip_info = parseOperationWithLLM(operation, prompt, schema);

        // Get track ID from context
        std::string track_id = getTrackIdFromContext(context);

        // Generate unique clip ID
        std::string clip_id = generateUniqueId();

        // Create clip result
        ClipResult clip_result;
        clip_result.clip_id = clip_id;
        clip_result.track_id = track_id;
        clip_result.start_bar = clip_info.value("start_bar", 1);
        clip_result.end_bar = clip_info.value("end_bar", clip_result.start_bar + 4);

        // Handle optional parameters
        if (clip_info.contains("start_time")) {
            clip_result.start_time = clip_info["start_time"];
        }
        if (clip_info.contains("duration")) {
            clip_result.duration = clip_info["duration"];
        }
        if (clip_info.contains("track_name")) {
            clip_result.track_name = clip_info["track_name"];
        }

        // Store clip for future reference
        clips_[clip_id] = clip_result;

        // Generate DAW command
        std::string daw_command = generateDawCommand(clip_result);

        // Create result JSON
        nlohmann::json result;
        result["id"] = clip_id;
        result["track_id"] = track_id;
        result["start_bar"] = clip_result.start_bar;
        result["end_bar"] = clip_result.end_bar;

        if (clip_result.start_time.has_value()) {
            result["start_time"] = clip_result.start_time.value();
        }
        if (clip_result.duration.has_value()) {
            result["duration"] = clip_result.duration.value();
        }
        if (!clip_result.track_name.empty()) {
            result["track_name"] = clip_result.track_name;
        }

        return AgentResponse(result, daw_command, context);

    } catch (const std::exception& e) {
        nlohmann::json error_result;
        error_result["error"] = "Error executing clip operation: " + std::string(e.what());
        return AgentResponse(error_result, "", context);
    }
}

std::vector<std::string> ClipAgent::getCapabilities() const {
    return {"clip", "region", "recording", "audio clip", "midi clip"};
}

std::optional<ClipResult> ClipAgent::getClipById(const std::string& clip_id) const {
    auto it = clips_.find(clip_id);
    if (it != clips_.end()) {
        return it->second;
    }
    return std::nullopt;
}

std::vector<ClipResult> ClipAgent::listClips() const {
    std::vector<ClipResult> result;
    for (const auto& pair : clips_) {
        result.push_back(pair.second);
    }
    return result;
}

std::string ClipAgent::generateDawCommand(const ClipResult& clip) const {
    std::ostringstream oss;
    oss << "clip(track:" << clip.track_id
        << ", start_bar:" << clip.start_bar
        << ", end_bar:" << clip.end_bar;

    if (clip.start_time.has_value()) {
        oss << ", start_time:" << clip.start_time.value();
    }
    if (clip.duration.has_value()) {
        oss << ", duration:" << clip.duration.value();
    }

    oss << ")";
    return oss.str();
}

std::string ClipAgent::generateDAWCommand(const nlohmann::json& result) const {
    std::ostringstream oss;
    oss << "clip(track:" << result.value("track_id", "unknown")
        << ", start_bar:" << result.value("start_bar", 1)
        << ", end_bar:" << result.value("end_bar", 5);

    if (result.contains("start_time")) {
        oss << ", start_time:" << result["start_time"];
    }
    if (result.contains("duration")) {
        oss << ", duration:" << result["duration"];
    }

    oss << ")";
    return oss.str();
}

std::string ClipAgent::getTrackIdFromContext(const nlohmann::json& context) const {
    // Try to get track ID from context parameters
    if (context.contains("track_id")) {
        return context["track_id"];
    }
    if (context.contains("track_daw_id")) {
        return context["track_daw_id"];
    }

    // Fallback to old context method
    if (context.contains("track") && context["track"].contains("id")) {
        return context["track"]["id"];
    }

    return "unknown";
}

} // namespace magda
